src/scheduler.ml

```diff
@@ -54,24 +156,42 @@ end = struct
     | Some x -> x
 
   let wait_unix () =
-    let pid, status = Unix.wait () in
-    (resolve_and_remove_job pid, status)
+    Unix.wait ()
 
   let wait =
     if Sys.win32 then
       wait_win32
     else
       wait_unix
+
+  let rec wait_to_chan chan =
+    Mutex.lock i_count_mtx;
+    if !i_count = 0 then
+      Condition.wait i_count_cv i_count_mtx;
+    i_count := !i_count - 1;
```

Comment:
If we decrement before the `Unix.wait`, then `i_count` doesn't exactly represent the number of child processes. For instance it means that we are going to call `Condition.signal` more often than necessary. To avoid this, we should decrement it after `Unix.wait` has returned, or to avoid taking the mutex twice in this loop, we can do it just after the first `Mutex.lock`, except on the first execution of this loop. In the end, I think it looks like this:


```ocaml
let rec wait_to_chan chan =
  if !i_count = 0 then ...
  Mutex.unlock i_count_mtx;
  let pid, status = wait () in
  Event....
  Mutex.lock ...
  i_count := !i_count - 1;
  wait_to_chan chan

let wait_to_chan chan =
  Mutex.lock i_count_mtx;
  wait_to_chan chan
```